pkg/nodeagent/registry/containerd_extractor.go

```diff
@@ -133,56 +134,122 @@ func unmountImage(ctx context.Context, snapshotter snapshots.Snapshotter, direct
 }
 
 // CopyFile copies a source file to destination file and sets the given permissions.
-func CopyFile(fs afero.Afero, sourceFile, destinationFile string, permissions os.FileMode) error {
-	sourceFileStat, err := fs.Stat(sourceFile)
+func CopyFile(fs afero.Afero, source, destination string, permissions os.FileMode) error {
+	if destinationFileStat, err := fs.Stat(destination); err == nil {
+		if !destinationFileStat.Mode().IsRegular() {
+			return fmt.Errorf("destination %q exists but is not a regular file", destination)
+		}
+	} else if !os.IsNotExist(err) {
+		return err
+	}
+
+	if err := fs.MkdirAll(path.Dir(destination), permissions); err != nil {
+		return fmt.Errorf("destination directory %q could not be created", path.Dir(destination))
+	}
+
+	tempDir, err := fs.TempDir(nodeagentv1alpha1.TempDir, "copy-image-")
 	if err != nil {
+		return fmt.Errorf("error creating temp directory: %w", err)
+	}
+
+	defer func() { utilruntime.HandleError(fs.Remove(tempDir)) }()
+
+	tmpFilePath := filepath.Join(tempDir, filepath.Base(destination))
+
+	if err := copyFileAndSync(fs, source, tmpFilePath); err != nil {
+		return err
+	}
+
+	if err := fs.Chmod(tmpFilePath, permissions); err != nil {
 		return err
 	}
 
+	return MoveFile(fs, tmpFilePath, destination)
+}
+
+func copyFileAndSync(fs afero.Afero, source, destination string) error {
+	sourceFileStat, err := fs.Stat(source)
+	if err != nil {
+		return err
+	}
 	if !sourceFileStat.Mode().IsRegular() {
-		return fmt.Errorf("source %q is not a regular file", sourceFile)
+		return fmt.Errorf("source %q is not a regular file", source)
 	}
 
-	if destinationFileStat, err := fs.Stat(destinationFile); err == nil {
+	if destinationFileStat, err := fs.Stat(destination); err == nil {
```

Comment:
as this doesn't use `Lstat` (do we really want to follow symlinks here?), it would be cleaner to check the file stats after opening the destination file. Then `dstFile.Stat()` is available and race condition free.

Reply:
I'm pretty sure that the file won't be a symlink in practice, because it is created by `gardener-node-agent` in the first place.

We are in a bit of a hurry now, because today is the last day where we could cherry-pick this PR in v1.90 where `gardener-node-agent` became GA.
Thus, I suggest to do these small improvements in a separate PR.

/unhold